test/kt/single_pass_scan.cpp

```diff
@@ -0,0 +1,181 @@
+    TestUtils::usm_data_transfer<_alloc_type, T> dt_output(q, size);
+
+    std::inclusive_scan(expected.begin(), expected.end(), expected.begin(), bin_op);
+
+    oneapi::dpl::experimental::kt::gpu::inclusive_scan(q, dt_input.get_data(), dt_input.get_data() + size,
+                                                  dt_output.get_data(), bin_op, param)
+        .wait();
+
+    std::vector<T> actual(size);
+    dt_output.retrieve_data(actual.begin());
+
+    std::string msg = "wrong results with USM, n: " + std::to_string(size);
+    EXPECT_EQ_N(expected.begin(), actual.begin(), size, msg.c_str());
+}
+
+template <typename T, typename BinOp, typename KernelParam>
+void
+test_sycl_iterators(sycl::queue q, std::size_t size, BinOp bin_op, KernelParam)
+{
+    constexpr oneapi::dpl::experimental::kt::kernel_param<KernelParam::data_per_workitem, KernelParam::workgroup_size>
+        param;
+#if LOG_TEST_INFO
+    std::cout << "\t\ttest_sycl_iterators<" << TypeInfo().name<T>() << ">(" << size << ");" << std::endl;
+#endif
+    std::vector<T> input(size);
+    std::vector<T> output(size);
+    generate_scan_data(input.data(), size, 42);
+    std::vector<T> ref(input);
+    std::inclusive_scan(std::begin(ref), std::end(ref), std::begin(ref), bin_op);
+    {
+        sycl::buffer<T> buf(input.data(), input.size());
+        sycl::buffer<T> buf_out(output.data(), output.size());
+        oneapi::dpl::experimental::kt::gpu::inclusive_scan(q, oneapi::dpl::begin(buf), oneapi::dpl::end(buf),
+                                                      oneapi::dpl::begin(buf_out), bin_op, param)
+            .wait();
+    }
+
+    std::string msg = "wrong results with oneapi::dpl::begin/end, n: " + std::to_string(size);
+    EXPECT_EQ_RANGES(ref, output, msg.c_str());
+}
+
+template <typename T, typename BinOp, typename KernelParam>
+void
+test_general_cases(sycl::queue q, std::size_t size, BinOp bin_op, KernelParam param)
+{
+    test_usm<T, sycl::usm::alloc::shared>(q, size, bin_op, param);
+    test_usm<T, sycl::usm::alloc::device>(q, size, bin_op, param);
+    test_sycl_iterators<T>(q, size, bin_op, param);
+#if _ENABLE_RANGES_TESTING
+    test_all_view<T>(q, size, bin_op, param);
+#endif
+}
+
+template <typename T, typename KernelParam>
+void
+test_all_cases(sycl::queue q, std::size_t size, KernelParam param)
+{
+    test_general_cases<T>(q, size, std::plus<T>{}, param);
+    test_general_cases<T>(q, size, std::multiplies<T>{}, param);
+}
+
+int
+main()
+{
+#if LOG_TEST_INFO
+    std::cout << "TEST_DATA_PER_WORK_ITEM : " << TEST_DATA_PER_WORK_ITEM << "\n"
+              << "TEST_WORK_GROUP_SIZE    : " << TEST_WORK_GROUP_SIZE << "\n"
+              << "TEST_TYPE               : " << TypeInfo().name<TEST_TYPE>() << std::endl;
+#endif
+
+    constexpr oneapi::dpl::experimental::kt::kernel_param<TEST_DATA_PER_WORK_ITEM, TEST_WORK_GROUP_SIZE> params;
+    auto q = TestUtils::get_test_queue();
+    try
+    {
+        for (auto size : scan_sizes)
```

Comment:
Optional:
Do we expect different problems to occur with different input data types?
I'm wondering with all the permutations, if we could get away with only running a subset of the input sizes on some input sequence data types for the sake of time.  For instance, we could always run all for USM device input data, but for USM shared, sycl_iterator, all_view, perhaps we could pick a subset of sizes (one single wg, one multi wg or something).

I think the same question could be asked of esimd radix sort.

Reply:
I suggest doing that after revising the API. We may end up in reducing the amount of possible imputs.